src/Installer/dnup/DnupSharedManifest.cs

```diff
@@ -0,0 +1,135 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.Json;
+using System.Threading;
+using Microsoft.Dotnet.Installation.Internal;
+
+namespace Microsoft.DotNet.Tools.Bootstrapper;
+
+internal class DnupSharedManifest : IDnupManifest
+{
+    private string ManifestPath => GetManifestPath();
+
+    public DnupSharedManifest(string? manifestPath = null)
+    {
+        _customManifestPath = manifestPath;
+        EnsureManifestExists();
+    }
+
+    private void EnsureManifestExists()
+    {
+        if (!File.Exists(ManifestPath))
+        {
+            Directory.CreateDirectory(Path.GetDirectoryName(ManifestPath)!);
+            File.WriteAllText(ManifestPath, JsonSerializer.Serialize(new List<DotnetInstall>(), DnupManifestJsonContext.Default.ListDotnetInstall));
+        }
+    }
+
+    private string? _customManifestPath;
+
+    private string GetManifestPath()
+    {
+        // Use explicitly provided path first (constructor argument)
+        if (!string.IsNullOrEmpty(_customManifestPath))
+        {
+            return _customManifestPath;
+        }
+
+        // Fall back to environment variable override
+        var overridePath = Environment.GetEnvironmentVariable("DOTNET_TESTHOOK_MANIFEST_PATH");
+        if (!string.IsNullOrEmpty(overridePath))
+        {
+            return overridePath;
+        }
+
+        // Default location
+        return Path.Combine(
+            Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
+            "dnup",
+            "dnup_manifest.json");
+    }
+
+    private void AssertHasFinalizationMutex()
+    {
+        // Instead of attempting to reacquire the named mutex (which can create race conditions
+        // or accidentally succeed when we *don't* hold it), rely on the thread-local tracking
+        // implemented in ScopedMutex. This ensures we only assert based on a lock we actually obtained.
+        if (!ScopedMutex.CurrentThreadHoldsMutex)
+        {
+            throw new InvalidOperationException("The dnup manifest was accessed without holding the installation state mutex.");
+        }
+    }
+
+    public IEnumerable<DotnetInstall> GetInstalledVersions(IInstallationValidator? validator = null)
+    {
+        AssertHasFinalizationMutex();
+        EnsureManifestExists();
+
+        var json = File.ReadAllText(ManifestPath);
```

Comment:
This reads the entire content of the file into memory. Did you consider using a `Stream` for deserialization? Intuition is that is going to be more efficient.

Reply:
That's a great idea. The manifest file will be very small, so I'm curious to see the numbers when we implement that. The manifest is very barebones - eventually we want a json schema file, caching, ability to upgrade / work with breaking changes to the manifest, manifest version metadata, etc. All of that is tracked in https://github.com/dotnet/sdk/issues/51099, and I'll add the `Stream` idea, thank you!